src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+            .filter(|txn| txn.hash.is_some())
+            .map(|txn| (txn.hash.as_ref().unwrap().clone(), txn.uid().clone()))
+            .collect();
+        let mut expiring_txns = BTreeMap::new();
+        let mut uids_awaiting_memos = HashMap::new();
+        for txn in store.iter() {
+            expiring_txns
+                .entry(txn.timeout())
+                .or_insert_with(HashSet::default)
+                .insert(txn.uid().clone());
+            for pending in &txn.pending_uids {
+                uids_awaiting_memos.insert(*pending, txn.uid().clone());
+            }
+        }
+        Ok(Self {
+            store,
+            txn_by_hash,
+            expiring_txns,
+            uids_awaiting_memos,
+        })
+    }
+
+    /// Iterate through the assets.
+    pub fn iter(&self) -> impl Iterator<Item = Transaction<L>> + '_ {
+        self.store.iter().cloned()
+    }
+
+    /// Get the asset by the code from the store.
+    pub fn get(&self, uid: &TransactionUID<L>) -> Result<Transaction<L>, KeystoreError<L>> {
+        Ok(self.store.load(uid)?)
+    }
+
+    /// Get a mutable asset editor by the code from the store.
+    pub fn get_mut(
+        &mut self,
+        uid: &TransactionUID<L>,
+    ) -> Result<TransactionEditor<'_, L>, KeystoreError<L>> {
+        let txn = self.get(uid)?;
+        Ok(TransactionEditor::new(&mut self.store, txn))
+    }
+
+    /// Get a Transaction with the id of a memo, will return the transaction which is awaiting the memo
+    pub fn with_memo_id(&self, id: u64) -> Result<Transaction<L>, KeystoreError<L>> {
+        let uid = self
+            .uids_awaiting_memos
+            .get(&id)
+            .ok_or(KeyValueStoreError::KeyNotFound)?;
+        Ok(self.get(uid).unwrap())
+    }
+
+    /// Same as with_memo_id but return a TransactionEditor
+    pub fn with_memo_id_mut(
+        &mut self,
+        id: u64,
+    ) -> Result<TransactionEditor<'_, L>, KeystoreError<L>> {
+        let txn = self.with_memo_id(id)?;
+        Ok(TransactionEditor::new(&mut self.store, txn))
+    }
+
+    /// Get a Transaction for a given TransactionHash
+    pub fn with_hash(&self, hash: &TransactionHash<L>) -> Result<Transaction<L>, KeystoreError<L>> {
+        let uid = self
+            .txn_by_hash
+            .get(hash)
+            .ok_or(KeyValueStoreError::KeyNotFound)?;
+        Ok(self.get(uid).unwrap())
+    }
+
+    /// Get a TransactionEditor for a given TransactionHash
+    pub fn with_hash_mut(
+        &mut self,
+        hash: &TransactionHash<L>,
+    ) -> Result<TransactionEditor<'_, L>, KeystoreError<L>> {
+        let txn = self.with_hash(hash)?;
+        Ok(TransactionEditor::new(&mut self.store, txn))
+    }
+
+    /// Get all the transactions timing out at the provided time.  There is no mutable way to get this
+    /// Collection because we should only be edditing one thing at a time
+    pub fn with_timeout(&self, timeout: u64) -> Result<Vec<Transaction<L>>, KeystoreError<L>> {
```

Comment:
This should probably return `Result<impl Iterator<Item = Transaction<L>>, KeystoreError<L>>`, so that we don't have to `collect()` in the implementation when the caller may only want to iterate